### ISSUE TITLE
Enhancement proposal: German commands for Turtleizer

### ISSUE THREAD

**Issue by codemanyak** [fesch/Structorizer.Desktop#490]
Obtained via mail (@andybru):

> Anbei mal mein Versuch, der so mit Structorizer zu funktionieren scheint, den Turtleizer auf 'deutsch' zu erlauben (sehr ähnlich der LibreOffice Ergänzung LibreLogo).
Zuerst hatte ich gehofft, es würde genügen, nur die Turtleizer.java anzupassen.
Das alles hat aber erst  dann funktioniert, als in der Datei TurtleBox.java der Bereich in den Zeilen 133 bis 154 ergänzt wurde.
Insgesamt scheint mir da etwas Redundanz in der "Sprach-Definition" zu existieren, aber ich habe es nicht weiter hinterfragt.
Vielleicht sind in einem zukünftigen Structorizer Release ja diese oder ähnliche Änderungen die Veröffentlichung wert.
Anbei ein pdf und die betreffenden Java-Dateien mit den Änderungen.
Getestet habe ich es mit der SchildkrötenHausDE.nsd ..

Translation of the proposal:

> I send you an attempt of mine to allow a Turtleizer in German (very similar to the LibreOffice enhancement LibreLogo) that seems to work with Structorizer.
At first I had hoped it might be sufficient to adapt file Turtleizer.java.
But then it only began to work when rows 133 through 154 were inserted in file TurtleBox.java.
In general it seems to me as if there were some redundancy in the API specification but I haven't scrutinized it any further.
Perhaps these or similar modifications might be worth to be published in some future release.
In the attachment there is a pdf and the related Java files with the modifications.
I tested it with the diagram SchildkrötenHausDE.nsd ..

[TurtleBox_und_Turtleizer_Aenderungen.zip](https://github.com/fesch/Structorizer.Desktop/files/1552073/TurtleBox_und_Turtleizer_Aenderungen.zip)

![schildkroetenhaus](https://user-images.githubusercontent.com/15326471/33891967-d130aa24-df57-11e7-8076-347a0ba6d6d0.png)

**Comment by codemanyak** [comment]
First of all: The apparent redundancy among TurtleBox.java and Turtleizer.java cannot be avoided whie a clean approach is desired. On the one hand (TurtleBox.java), there is the plugin interface for the internal interpeter (based on the Java interface class DiagramController) and assuming case-ignorant API use (therefore all routine keys are lower-case ids combined with argument number in the look-up table), whereas on the other hand Turtleizer.java is just an external adapter for Java code exported by Structorizer. The mapping table located there enables the Java generator to replace Turtleizer routine identifiers (case-ignorant) by the actual TurtleBox method calls, circumventing the inperformant interpreter plugs. Despite of the geater simlicity of the latter (only maps Structorizer command name to Java method name) both tables correspond.

But now to your proposal. I don't deem a hard-coded integration of German aliases into the product a good paradigm because then why don't we introduce hard-coded aliases for other languages, too? And this would not only refer to the natural user languages, but as well to alternative Turtleizer emulation libraries, e.g. for the export to C++. This would not only multiply the efforts for documentation but might also provoke unpleasant name conflicts with user-defined routine names in existing diagrams.

So I would rather propose an oportunity for the user to specify individual aliases for Turtleizer commands, which would then be resolved before (or n) execution. Principally, this would only be a convenience since every user might define their iown set of simple adapter diagrams for each for a Turtleizer command, and then to store them together in an .arrz file, but this would have some disadvantages, chiefly tediousness and low performance.

But once having the API tables in TurtleBox, we could easily derive a configuration dialog offering the definition of aliases for all inteface routines. During export to some programming language, corresponding adapter functions or methods would indeed have to be generated automatically, e.g. to Java:
```
private void vorwärts(double pixel) {
     forward(pixel);
}
```
This could be implemented for all export languages in an analogous way and would stay flexible in any aspect. The mechanism would also be open for potential other controllable plugins.

**Comment by codemanyak** [comment]
After some consideration and imagining the rather uncomfortable consequence having to offer a refactoring of used Turtle command aliases (similar to the instruction keywords now), I would rather propose an even more attractive approach now:

A different configurable solution, also preserving portability of the diagrams without significant overhead, would not store the alias names in the diagrams but might simply use the aliases for just the display of the diagram while internally "actual" routine names would be stored like they are now (similar to the replacement of assignment operator `<-` or some combined comparison operators by the left arrow symbol ← or ligatures). On editing elements they should also appear in the text area but would be translated back to the genuine names on storing. This means that no modification of Executor or cde generators would be necessary at all.
And any user having configured individual aliases would see Turtle commands in loaded diagrams in terms of her/his own favourite identifiers.

**Comment by codemanyak** [comment]
TurtleBox slightly revised to offer a light-weight standard constructor (to allow API retrieval without materializing the Turtleizer window), thus allowing to develop a generic alias preferences dialog.

**Comment by codemanyak** [comment]
The new preferences menu for the alias definition is ready, it works on display and editing (bugfix branch):
![diagramcontroleraliases3 27-05](https://user-images.githubusercontent.com/15326471/35202240-cb63b386-ff21-11e7-9198-55e4208d607d.png)
If the replacement is enabled (the checkbox in the dialog) then the Turtleizer routines will be presented with their alias names in the diagram instead (as specified). While variable highlighting is active, the alias names will appear underlined:
![drawmaeander-4](https://user-images.githubusercontent.com/15326471/35202337-4e1cf80a-ff22-11e7-9029-464df0efb082.png)
In the element editor it doesn't make a difference whether the real routine name or the alias is entered, the result is identical: The true routine name is stored while the alias is shown. For an existing text the editor will also show the alias.
The redundant routine configuration in the Java adapter class `Turtleizer´ is no longer necessary for the Java generator, I managed to derive a light-weight `TurtleBox´ instance, such that the API can now be retrieved directly there without unwilingly opening the Turtleizer window.

The Find & Replace dialog still works on the genuine routine names, possible error messages in Executor, too. As soon as this is fixed, I'll push Structorizer version 3.27-05.
